master/internal/hpimportance/manager.go

```diff
@@ -0,0 +1,284 @@
+func NewManager(db *db.PgDB) actor.Actor {
+	return &manager{db, make(map[int]stateRecord)}
+}
+
+func (m *manager) Receive(ctx *actor.Context) error {
+	switch msg := ctx.Message().(type) {
+	case actor.PreStart:
+		// TODO: fetch any pending or in_progress tasks from the DB and trigger them
+	case actor.PostStop:
+		// Do nothing
+	case actor.ChildFailed:
+		ctx.Log().Warnf("hyperparameter importance worker failed: %+v", msg)
+	case actor.ChildStopped:
+		// Do nothing - it'll respawn next time a request is received
+	case ExperimentCompleted:
+		m.triggerDefaultWork(ctx, msg.ID)
+	case ExperimentCreated:
+		m.state[msg.ID] = stateRecord{
+			lastResult: time.Now(),
+		}
+	case ExperimentPaused:
+		m.triggerDefaultWork(ctx, msg.ID)
+	case ExperimentProgress:
+		var state stateRecord
+		var ok bool
+		if state, ok = m.state[msg.ID]; !ok {
+			state = stateRecord{}
+		}
+		if msg.Progress-state.lastProgress > minPercent &&
+			time.Since(state.lastResult) > minPause {
+			m.triggerDefaultWork(ctx, msg.ID)
+		}
+	case workStarted:
+		hpi, err := m.db.GetHPImportance(msg.experimentID)
+		if err != nil {
+			ctx.Log().Errorf("error retrieving hyperparameter importance state: %s", err.Error())
+			return nil
+		}
+		metricData := getMetricHPImportance(hpi, msg.metricName, msg.metricType)
+		metricData.Status = InProgress
+		setMetricHPImportance(&hpi, metricData, msg.metricName, msg.metricType)
+		err = m.db.SetHPImportance(msg.experimentID, hpi)
+		if err != nil {
+			ctx.Log().Errorf("error writing hyperparameter importance state: %s", err.Error())
+			return nil
+		}
+	case workFailed:
+		hpi, err := m.db.GetHPImportance(msg.experimentID)
+		if err != nil {
+			ctx.Log().Errorf("error retrieving hyperparameter importance state: %s", err.Error())
+			return nil
+		}
+		metricData := getMetricHPImportance(hpi, msg.metricName, msg.metricType)
+		metricData.Status = Failed
+		metricData.Error = msg.err
+		setMetricHPImportance(&hpi, metricData, msg.metricName, msg.metricType)
+		err = m.db.SetHPImportance(msg.experimentID, hpi)
+		if err != nil {
+			ctx.Log().Errorf("error writing hyperparameter importance state: %s", err.Error())
+			return nil
+		}
+	case workCompleted:
+		m.state[msg.experimentID] = stateRecord{
+			lastResult:   time.Now(),
+			lastProgress: m.state[msg.experimentID].lastProgress,
+		}
+		hpi, err := m.db.GetHPImportance(msg.experimentID)
+		if err != nil {
+			ctx.Log().Errorf("error retrieving hyperparameter importance state: %s", err.Error())
+			return nil
+		}
+		metricData := getMetricHPImportance(hpi, msg.metricName, msg.metricType)
+		metricData.ExperimentProgress = msg.progress
+		metricData.HpImportance = msg.results
+		switch metricData.Status {
+		case Pending:
+			// Do nothing - this means another startWork message was already sent
+		case InProgress:
+			metricData.Status = Complete
+		default:
+			ctx.Log().Warnf("work was completed for a metric with an unexpected state")
```

Comment:
if this is invalid, should we probably just bail?

Reply:
I don't think so. It would mean there's either a bug, or some transition in the state machine I didn't anticipate. But I'm struggling to think of one where the right course of action is to discard the results we just computed. It doesn't mean the results are useless or wrong - they *might* just be about to get overwritten because a parallel task is also working on it. But there's still no big side effect. It's just, if there's some other transition here, and it shows up in someone's logs one day? I'd be very curious to look at it.

Reply:
> It would mean there's either a minor bug, or some transition in the state machine I didn't anticipate.

My thought generally is if we've left the path of intended behaviors, we should just fail because i usually havent put any energy into imagining what can happen otherwise.

Reply:
I thought through this and even if such an invalid transition slips through, it seems like the right thing to do is always to just move to the latest state anyway, unless any state transition is missing enough to cause other functional problems, which would be a separate bug in it's own right. Just removed these.